services/processing/merging.py

```diff
@@ -69,17 +81,80 @@ def update_uploads(db_session: DbSession, merge_result: MergeResult):
             db_session, report_id, merge_result.deleted_sessions
         )
 
-    # then, update all the sessions that have been merged
-    for upload_id, session_id in merge_result.session_mapping.items():
-        update = {
-            Upload.state_id: UploadState.PROCESSED.db_id,
-            Upload.state: "processed",
-            Upload.order_number: session_id,
-        }
-        db_session.query(Upload).filter(Upload.id_ == upload_id).update(update)
+    precision: int = read_yaml_field(commit_yaml, ("coverage", "precision"), 2)
+    rounding: str = read_yaml_field(commit_yaml, ("coverage", "round"), "nearest")
+    make_totals = functools.partial(make_upload_totals, precision, rounding)
+
+    reports = {ir.upload_id: ir.report for ir in intermediate_reports}
+
+    # then, update all the `Upload`s with their state, and the final `order_number`,
+    # as well as add a `UploadLevelTotals` or `UploadError`s where appropriate.
+    all_errors: list[UploadError] = []
+    all_totals: list[dict] = []
+    all_upload_updates: list[dict] = []
+    for result in processing_results:
+        upload_id = result["upload_id"]
+
+        if result["successful"]:
+            update = {
+                "state_id": UploadState.PROCESSED.db_id,
+                "state": "processed",
+            }
+            report = reports.get(upload_id)
+            if report is not None:
+                all_totals.append(make_totals(upload_id, report.totals))
+        elif result["error"]:
+            update = {
+                "state_id": UploadState.ERROR.db_id,
+                "state": "error",
+            }
+            error = UploadError(
+                upload_id=upload_id,
+                error_code=result["error"]["code"],
+                error_params=result["error"]["params"],
+            )
+            all_errors.append(error)
+
+        update["id_"] = upload_id
+        order_number = merge_result.session_mapping.get(upload_id)
+        update["order_number"] = order_number
+        all_upload_updates.append(update)
+
+    db_session.bulk_update_mappings(Upload, all_upload_updates)
+    db_session.bulk_save_objects(all_errors)
+
+    # the `UploadLevelTotals` have a unique constraint for the `upload`,
+    # so we have to use a manual `insert` statement:
+    stmt = (
+        insert(UploadLevelTotals.__table__).values(all_totals).on_conflict_do_nothing()
```

Comment:
I don’t expect any conflicts here. In the very rare case that the task reruns, this code is not idempotent in that case and would error.
This is just a precaution for that rare case.
You might be right that in case of reruns, we would also double-insert the errors as well, which do not have such a unique constraint, and we would indeed end up with duplicated entries.

But as I said, I do not expect this case to ever be hit in reality (famous last words and all :-D)